keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1349 @@
+ensures they are formatted “correctly”. The only meaning of “correct” is “block
+style”, and it corrupts KYAML input today (aside: it also does not handle
+multi-documents).  If we intend to convert examples, we need to fix this.
+
+We propose to add flags to it which specify to use “conventional” YAML or KYAML
+output, and set the default to KYAML (or to change the parent script to always
+specify KYAML).
+
+This tooling will preserve comments.
+
+#### How?
+
+Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While
+KYAML is an output format, our tooling goals require that we get access to a
+YAML AST, which may require some dirty tricks.  If we ever change YAML
+libraries (e.g.
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we may
+need different dirty tricks.
+
+Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST,
+which we can render as KYAML.  Neither, however, offers a way to encode an
+arbitrary Go object as the equivalent YAML AST.  To do that we may have to
+encode the object to plain YAML, then decode it to AST, then render it as
+KYAML.  That extra conversion is ugly, but this is not generally
+performance-sensitive code and is not run on the server.
+
+[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml) offers an
+API which appears to do what we need in both directions.
+
+### Header
+
+At the top of each KYAML document (or multi-document) will be the following header:
+
+```
+#!yaml
+---
+```
+
+This disambiguates it from JSON, since both KYAML and JSON begin with an
+opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML
+without the header, but it is needed for versions before that.
+
+### Per-type rendering
+
+Integers and floats are rendered as their natural numeric representation.
+
+Simple strings are rendered as double-quoted and escaped. Multi-line strings
+require special consideration, see below.
+
+Booleans are rendered as `true` or `false`.
+
+Structs are rendered with curly-braces, and each field on a line of its own.
+Field names are not quoted unless they are not obviously strings (which
+includes prefixed label keys) or are one of the known type-ambiguous words
+(e.g. `no`).  Field values are rendered as per their type.  Embedded structs
+will be rendered as inline fields.
+
+KYAML does not guarantee that fields will be rendered in the order they are
+defined in the struct.  Implementations may preserve field order, but `kubectl`
+itself does not guarantee this (it converts to a map internally, which loses
+the type schema), so the order may be different.
+
+Fields which have a `json` struct tag will be rendered according to that tag:
+  * the tag’s field name will be used
+  * if the tag’s name is “-”, the field will be skipped
+  * if the tag specifies “omitempty” it will be respected
+  * if the tag specifies “omitzero” it will be respected
```

Comment:
Ordering (of order-independent bits) and formatting can differ, but the content and structure of the output must be identical to `-o json` and `-o yaml`, right?

Rendering as json using the standard json encoder and then transforming into kyaml seems like the most bullet-proof way to accomplish this

Reply:
kubectl turns everything into `unstructured` long before the final `Printer` sees the data.  Unstructured implements `json.Marshaler`, which my POC supports directly, so everything gets bounced thru JSON.

I suppose a different approach would be to just bounce EVERYTHING thru JSON, which means I don't have to implement the tag logic (already done, but still).  It means we ALWAYS pessimize the field ordering, no matter where you apply this code.  And it would always be as slow as possible.

I could go either way.  It's EFFECTIVELY what is happening already in kubectl.

Reply:
I think the generic kubectl commands all use unstructured, but some of the commands do use typed objects as output. The more we can do to auto-inherit JSON serialization semantics and make this *just* do yaml reflowing, the better

Reply:
> kubectl turns everything into `unstructured` long before the final Printer sees the data. 

Random thought (unrelated to this KEP): what if kubectl used something else than `unstructured` which implements the same interfaces (for example, meta data access) but preserves field ordering? I'm tempted to try it out...

Reply:
Probably, but it would be a major breaking change of our API. That it's a map is exposed: https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured

I suppose someone noticed that this might not be a good idea:
> TODO: make the serialization part of this type distinct from the field accessors.

Reply:
You can do some really surprising things in YAML map keys that we should be aware of but may choose to ignore here. YAML only requires that keys be unique. Ideally we want KYAML to be fairly reasonable beyond just kubectl printing though as we get the most benefit when it's generally usable in the ecosystem to prefer formatting this way.

we somewhat address this already with the "sometimes quote keys"